utilities/configure_scripts/configure_dealii.sh

```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
+## ---------------------------------------------------------------------
+##
+## PartExa - A Particle Library for the Exa-Scale
+##
+## Copyright (C) 2022 by the PartExa authors
+##
+## This program is free software: you can redistribute it and/or modify
+## it under the terms of the GNU General Public License as published by
+## the Free Software Foundation, either version 3 of the License, or
+## (at your option) any later version.
+##
+## This program is distributed in the hope that it will be useful,
+## but WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with this program. If not, see <https://www.gnu.org/licenses/>.
+##
+## ---------------------------------------------------------------------
+
+# You may want to deactivate this line
+rm -rf CMakeFiles/ CMakeCache.txt
+
+# It is recommended to copy this file to your deal.II build directory via
+# ```bash
+# cd <dealii_build>
+# cp <partexa_source>/utilities/configure_scripts/configure_partexa.sh .
+# ```
+
+# Having copied the file, you need to adjust the paths specified below
+P4EST=/path/to/p4est
+DEAL=/path/to/dealii
+
+cmake \
+    -D CMAKE_BUILD_TYPE="DebugRelease" \
+    -D CMAKE_CXX_FLAGS="-std=c++17 -march=native -Wno-array-bounds -Wno-literal-suffix -pthread" \
+    -D DEAL_II_CXX_FLAGS_RELEASE="-O3" \
+    -D DEAL_II_CXX_FLAGS_DEBUG="-Og" \
+    -D CMAKE_C_FLAGS="-march=native -Wno-array-bounds" \
+    -D DEAL_II_WITH_MPI:BOOL="ON" \
+    -D DEAL_II_LINKER_FLAGS="-lpthread" \
+    -D DEAL_II_WITH_64BIT_INDICES="ON" \
+    -D DEAL_II_WITH_P4EST="ON" \
+    -D P4EST_DIR="$P4EST" \
+    -D DEAL_II_COMPONENT_DOCUMENTATION="OFF" \
+    -D DEAL_II_COMPONENT_EXAMPLES="OFF" \
+    $DEAL
```

Comment:
Let's go with the default flags here. We can adapt the flags later when needed and when it is clear what we need.
```suggestion
cmake \
    -D CMAKE_BUILD_TYPE="DebugRelease" \
    -D CMAKE_INSTALL_PREFIX="$DEAL_INSTALL" \
    -D DEAL_II_WITH_MPI="ON" \
    -D DEAL_II_WITH_P4EST="ON" \
    -D P4EST_DIR="$P4EST" \
    $DEAL
```

Reply:
Right now, I also like to go with the default flags here as proposed in the deal.II documentation and also given in our Sphinx documentation. Later on we can still decide to specialize the flags.

Reply:
I forgot to add: I will take care to adapt the Sphinx Documentation to account for the new config scripts after this PR is merged.

Reply:
can you refer to a page that lists what the default flags are?

Reply:
I don't get why you want to remove e.g. something like `DEAL_II_WITH_64BIT_INDICES` (without even discussing the need of those items).

What I can say is: At least in former times, strange errors occurred for large parallel computations when forgetting this flag as far as I know, see https://www.dealii.org/developer/doxygen/deal.II/namespacetypes.html#a3543786f7dc7c57385fc923a6afd5917 (search for global_dof_index), and e.g. the source code
```c++
#ifdef DEAL_II_WITH_64BIT_INDICES
  using global_cell_index = uint64_t;
#else
  using global_dof_index = unsigned int;
#endif
```

Reply:
Regarding the optimization flags, I think it would be good to consult also @kronbichler, but not just remove them ...

Reply:
Regarding the 64 bit indices: They are needed in case the number of unknowns (or cells in the mesh, or particles) accumulated over all processes goes beyond 2^32 (for some packages, the limit is 2^31 because they use signed integers). It is possible to work with the usual integers (without enabling the 64 bit flag in the deal.II configuration) if one is sure that one stays below that threshold. One can always switch to 64 bit numbers later, but it might happen that one has used the wrong type somewhere (under the assumption that `global_dof_index` is `unsigned int`, one might have written `unsigned int` as the type of a vector, where it should have been `global_dof_index`). Those errors are relatively easy to find once one switches to 64 bit numbers as it is a clear compile error. The more subtle thing that happens is wrong run time behavior because one creates local variables `unsigned int` that should have been 64 bit numbers (narrowing type). But these can also happen when one develops in 64 bit integers (and forgets to test more than 4B unknowns).

So in summary, I would not say that underlying mechanism is "strange" as the outcome is what one should expect when unsigned integer overflow (i.e., wrapping around of numbers) occurs: The results will make no sense any more as the identification of unknowns fails, and the error can pop up at "strange" locations, once one goes beyond the threshold of 4B unknowns.

Regarding the optimization flags: It is fine to work with the defaults if one does not care too much about special features (newer C++ standards) or performance. To change to something else might or might not give better results, and to make a selection requires careful benchmarking (as has been done in the past for certain problems of interest, which has resulted in the choices made e.g. by ExaDG). Of course, choosing `-O3` in release mode might make sense and can be expected reasonable for many applications, but it is clearly not a prerequisite at this stage. What flags get selected depends on the system at hand, the compiler, and settings given by the user. The flags used can be seen in the `detailed.log` file in the build fold of deal.II. The relevant part for the configuration that I'm using most often (Linux, clang-13 compiler, AMD Zen2 processor) is:
```
#        DEAL_II_CXX_FLAGS:            -pedantic -fPIC -Wall -Wextra -Wmissing-braces -Woverloaded-virtual -Wpointer-arith -Wsign-compare -Wsuggest-override -Wswitch -Wsynth -Wwrite-strings -Wno-deprecated-declarations -Wno-psabi -Wfloat-conversion -Qunused-arguments -Wno-unsupported-friend -Wno-undefined-var-template -openmp-simd -march=native -mfma -ffp-contract=fast -std=c++17
#        DEAL_II_CXX_FLAGS_RELEASE:    -O2 -funroll-loops -fstrict-aliasing -Wno-unused-local-typedefs -O3
#        DEAL_II_CXX_FLAGS_DEBUG:      -Og -ggdb -Wa,--compress-debug-sections
```
which is influenced by adding `-march=native` (in order to optimize the code for my hardware; in the optimized matrix-free paths this can be a factor 3 in performance on my machine or factor 5 on the best machines with AVX-512, while many other parts see much less of an improvement here) and `-std=c++17` to get a newer C++ standard, as well as adding `-O3` in the release compile flags. (Note that the last selection of -Ox takes precedence, so deal.II's default -O2 gets overriden by my `-O3` setting.)

Reply:
In any case, I believe someone who encounters the problem with int overflow (may it be called strange or not), will not necessarily know what to do or how to solve it in my opinion. This is why I suggested to include this parameter already now in the configure script, since it only helps but does not hurt.

@kronbichler What I did not understand is whether you would argue for "OFF" as default argument, in order to reduce memory requirements? I would argue that this "optimization" is not essential given that we will also store "a lot of" other data per particle for example (positions, velocities, etc. in double precision), so that we could just activate 64 bit indices by default? What do you think?

Reply:
I am split on this topic: I agree with your assessment that it can avoid some errors. Yet, I only enable 64 bit numbers on big machines (more than some dozen of nodes, where I regularly address problems with >2B unknowns) because it morally feels wrong in terms of waste of memory: `DoFHandler` consumes twice the memory with 64 bit numbers, all sparse matrices consume twice the memory for the index part (column indices stored as 64 bit integers rather than 32 bits), and a few more places. (I do not use big sparse matrices often, but when I do I prefer the faster option. Now that is clearly a shortcoming in the classes that could/should be addressed, but I'm not inclined on addressing it.) `MatrixFree` and some infrastructure around it are among the few parts of deal.II where the index storage type in local indices uses a narrower type (always 32 bit integers) also when the overall code is 64 bit, as that part is performance-sensitive enough to care. That said, this is a special case because there are regular tests that check both the 32 bit and 64 bit path.

For particles I agree that it will not be more than a few percent of the memory, so my feeling of "waste" is indeed highly subjective. Either way, I would not feel that disabling 64 bit integers for most developers would lead to difficult-to-address problems later, as long as there is knowledge that care needs to be taken and that the ambition is to switch to 64 bit numbers for big computations.

Reply:
To answer your question first, you can find what I stated the default configuration of deal.II here [2.2 Configuring and building the library](https://dealii.org/developer/readme.html). Additionally, we added the configure flags for p4est (see https://dealii.org/developer/external-libs/p4est.html) as it is mandatory when using e.g. the deal.II `particlehandler`. This configuration is what is also currently documented in the [Sphinx documentation](https://partexa.github.io/PartExa-Sphinx/installation/dependencies.html). For the sake of completeness: The deal.II documentation also gives a brief list of advanced [3. Configuration options](https://www.dealii.org/developer/users/cmake_dealii.html).

I'd like to name the reasons why I would prefer to keep this simple configuration for now. @ppraegla and me just recently dived into the new world of using deal.II for a user project. So that said, I have the impression that for us and the project it is best to gradually evolve the configuration during the course of code development. Right now, as we are still at the beginning of the project with zero lines of code I would not feel comfortable with specifying the configuration fine-grained at an early stage.

In my opinion, this PR should be about introducing the configure scripts only and should not contain a change in the configuration itself. The conversation between you and @kronbichler about configuration/optimization flags showed me that this is a new topic that needs discussion in a separate issue first.